File: _Uncategorized/452_MinNumOfArrowsToBurstBalloons.cs
// 452. Minimum Number of Arrows to Burst Balloons
// Tags: greedy, sorting, visual
// 

public class Solution 
{
    public int FindMinArrowShots(int[][] points) 
    {
        // Edge case.
        if(points == null 
            || points.Length == 0
            || points[0] == null
            || points[0].Length == 0)
        {
            return -1;
        }

        // So as along as all the constraints are fulfilled, we will always need 
        // to make at least 1 shot to burst all of the balloons.
        int result = 1;

        // Sort array from x_0 ascending.
        // i.e. sort by our first index in the sub-array.
        Array.Sort(points, (x0, x1) => { return x0[0] - x1[0]; });

        // foreach(var p in points)
        // {
        //     Console.WriteLine($"[{p[0]}, {p[1]}]");
        // }

        // Find range, continue or increment.
        int lowerBound = Int32.MinValue;
        int upperBound = Int32.MaxValue;
        foreach(var p in points)
        {
            // Still in range. Update bounds to smallest, available range.
            if(lowerBound <= p[1] && upperBound >= p[0])
            {
                lowerBound = Math.Max(lowerBound, p[0]);
                upperBound = Math.Min(upperBound, p[1]);
            }
            // Out of range. Need another shot and to reset the bounds.
            else
            {
                result += 1;
                lowerBound = p[0];
                upperBound = p[1];
            }
        }

        return result;
    }
}

File: _Uncategorized/452_MinNumOfArrowsToBurstBalloons.rs
// 452. Minimum Number of Arrows to Burst Balloons
// Tags: greedy, sorting, visual
// 

use std::cmp;
impl Solution {
    pub fn find_min_arrow_shots(mut points: Vec<Vec<i32>>) -> i32 {
        let mut result = 1;

        // Sort array from x_0 ascending.
        // i.e. sort by our first index in the sub-array.
        points.sort_by(|x0, x1| x0.cmp(x1));

        // Find range, continue or increment.
        let mut lower_bound: i32 = i32::MIN;
        let mut upper_bound: i32 = i32::MAX;

        for p in points.iter() {
            if lower_bound <= p[1] && upper_bound >= p[0] {
                lower_bound = cmp::max(lower_bound, p[0]);
                upper_bound = cmp::min(upper_bound, p[1]);
            }
            else {
                result += 1;
                lower_bound = p[0];
                upper_bound = p[1];
            }
        }

        result
    }
}